modules/images/image-loading-optimization/detect.js

```diff
@@ -0,0 +1,279 @@
+	const adminBar =
+		/** @type {?HTMLDivElement} */ doc.getElementById( 'wpadminbar' );
+
+	// We need to capture the original elements and their breadcrumbs as early as possible in case JavaScript is
+	// mutating the DOM from the original HTML rendered by the server, in which case the breadcrumbs obtained from the
+	// client will no longer be valid on the server. As such, the results are stored in an array and not any live list.
+	const breadcrumbedImages = doc.body.querySelectorAll( 'img' );
+
+	// We do the same for elements with background images which are not data: URLs.
+	const breadcrumbedElementsWithBackgrounds = Array.from(
+		doc.body.querySelectorAll( '[style*="background"]' )
+	).filter( ( /** @type {Element} */ el ) =>
+		/url\(\s*['"](?!=data:)/.test( el.style.backgroundImage )
+	);
+
+	/** @type {Map<Element, Breadcrumb[]>} */
+	const breadcrumbedElementsMap = new Map(
+		[ ...breadcrumbedImages, ...breadcrumbedElementsWithBackgrounds ].map(
+			( element ) => [ element, getBreadcrumbs( element ) ]
+		)
+	);
+
+	// Ensure the DOM is loaded (although it surely already is since we're executing in a module).
+	await new Promise( ( resolve ) => {
+		if ( doc.readyState !== 'loading' ) {
+			resolve();
+		} else {
+			doc.addEventListener( 'DOMContentLoaded', resolve, { once: true } );
+		}
+	} );
+
+	/** @type {IntersectionObserverEntry[]} */
+	const elementIntersections = [];
+
+	/** @type {?IntersectionObserver} */
+	let intersectionObserver;
+
+	function disconnectIntersectionObserver() {
+		if ( intersectionObserver instanceof IntersectionObserver ) {
+			intersectionObserver.disconnect();
+			win.removeEventListener( 'scroll', disconnectIntersectionObserver ); // Clean up, even though this is registered with once:true.
+		}
+	}
+
+	// Wait for the intersection observer to report back on the initially-visible elements.
+	// Note that the first callback will include _all_ observed entries per <https://github.com/w3c/IntersectionObserver/issues/476>.
+	if ( breadcrumbedElementsMap.size > 0 ) {
+		await new Promise( ( resolve ) => {
+			intersectionObserver = new IntersectionObserver(
+				( entries ) => {
+					for ( const entry of entries ) {
+						if ( entry.isIntersecting ) {
+							elementIntersections.push( entry );
+						}
+					}
+					resolve();
+				},
+				{
+					root: null, // To watch for intersection relative to the device's viewport.
+					threshold: 0.0, // As soon as even one pixel is visible.
+				}
+			);
+
+			for ( const element of breadcrumbedElementsMap.keys() ) {
+				if ( ! adminBar || ! adminBar.contains( element ) ) {
+					intersectionObserver.observe( element );
+				}
+			}
+		} );
+
+		// Stop observing as soon as the page scrolls since we only want initial-viewport elements.
+		win.addEventListener( 'scroll', disconnectIntersectionObserver, {
+			once: true,
+			passive: true,
+		} );
+	}
+
+	// TODO: Use a local copy of web-vitals.
```

Comment:
Let me know if you'd like help with that

Reply:
This todo isn't to be done in this PR. I'm working on this in https://github.com/WordPress/performance/pull/878.

Reply:
Oops. I thought you commented on a different todo :)

Reply:
I was presuming that there would be a JS build process introduced to the plugin as part of #864 and so we could incorporate the use of a local copy of web-vitals as part of that. But actually it looks like it's not going to involve JS. If you think a JS build process could be incorporated as part of the PR, I'd appreciate your help. But otherwise I think this can be tackled in a subsequent PR.

Reply:
@westonruter Just looking through these changes now, I think it would be good to open a separate issue/PR for introducing a build process for JS. This can be against the feature branch because no other module needs it so far, and I don't anticipate any new module being introduced in the near future where it would be needed. So we can do that as part of this one. Let's just make sure the approach could be used by other modules too.